jib-core/src/main/java/com/google/cloud/tools/jib/hash/Digests.java

```diff
@@ -0,0 +1,102 @@
+// TODO: BlobDescriptor is merely a tuple of (size, digest). Rename BlobDescriptor to something
+// more general.
+public class Digests {
+
+  public static DescriptorDigest computeJsonDigest(JsonTemplate template) throws IOException {
+    return computeDigest(template, ByteStreams.nullOutputStream()).getDigest();
+  }
+
+  public static DescriptorDigest computeJsonDigest(List<? extends JsonTemplate> templates)
+      throws IOException {
+    WritableContents contents = contentsOut -> JsonTemplateMapper.writeTo(templates, contentsOut);
+    return computeDigest(contents, ByteStreams.nullOutputStream()).getDigest();
+  }
+
+  public static BlobDescriptor computeDigest(JsonTemplate template, OutputStream optionalOutStream)
+      throws IOException {
+    WritableContents contents = contentsOut -> JsonTemplateMapper.writeTo(template, contentsOut);
+    return computeDigest(contents, optionalOutStream);
+  }
+
+  public static BlobDescriptor computeDigest(InputStream inStream) throws IOException {
+    return computeDigest(inStream, ByteStreams.nullOutputStream());
+  }
+
+  /**
+   * Computes the digest by consuming the contents.
+   *
+   * @param contents the contents for which the digest is computed
+   * @return computed digest and bytes consumed
+   * @throws IOException if reading fails
+   */
+  public static BlobDescriptor computeDigest(WritableContents contents) throws IOException {
+    return computeDigest(contents, ByteStreams.nullOutputStream());
+  }
+
+  /**
+   * Computes the digest by consuming the contents of an {@link InputStream} and optionally copying
+   * it to an {@link OutputStream}. Returns the computed digest along with the size of the bytes
+   * consumed to compute the digest. Does not close either stream.
+   *
+   * @param inStream the stream to read the contents from
+   * @param optionalOutStream the stream to which the contents are copied
+   * @return computed digest and bytes consumed
+   * @throws IOException if reading from or writing fails
+   */
+  public static BlobDescriptor computeDigest(InputStream inStream, OutputStream optionalOutStream)
+      throws IOException {
+    WritableContents contents = contentsOut -> ByteStreams.copy(inStream, contentsOut);
+    return computeDigest(contents, optionalOutStream);
+  }
+
+  /**
+   * Computes the digest by consuming the contents and optionally copying it to an {@link
+   * OutputStream}. Returns the computed digest along with the size of the bytes consumed to compute
+   * the digest. Does not close the stream.
+   *
+   * @param contents the contents to compute digest for
+   * @param optionalOutStream the stream to which the contents are copied
+   * @return computed digest and bytes consumed
+   * @throws IOException if reading from or writing fails
+   */
+  public static BlobDescriptor computeDigest(
+      WritableContents contents, OutputStream optionalOutStream) throws IOException {
```

Comment:
Just a couple questions about `optionalOutStream`
- is it truly optional? Even calls not using it still pass in a nulloutputstream.
- should we use `OutputStream` instead of `OutStream`

Reply:
The name "optional" is there only because of the name "computeDigest" doesn't convey the message that it can additionally copy `WritableContents` into the out stream while computing the digest. Like, from the point of a user just trying to compute a digest, they won't get why they have to pass an output stream for what.

Maybe I should just rename the method to "computeDigestAndCopy", as I said in the meeting? WDYT?

For `outStream`, I don't think anyone can confuse that with something else than a stream for output, but if you are averse to that, I can go with `outputStream`.

Reply:
It's not so much confusion as much as consistency for `outStream` vs `outputStream`, so not a big deal.

I think we can actually deal with the `optionalOutStream` in a follow up PR.